src/cowboy_protocol.erl

```diff
@@ -466,27 +476,60 @@ next_request(Req, State=#state{req_keepalive=Keepalive, timeout=Timeout},
 			%% Flush the resp_sent message before moving on.
 			if HandlerRes =:= ok, Buffer =/= close ->
 					receive {cowboy_req, resp_sent} -> ok after 0 -> ok end,
-					?MODULE:parse_request(Buffer,
+					next_request(Buffer,
 						State#state{req_keepalive=Keepalive + 1,
-						until=until(Timeout)}, 0);
+						until=until(Timeout)});
 				true ->
-					terminate(State)
+					terminate(normal, State)
 			end
 	end.
 
--spec error_terminate(cowboy:http_status(), #state{}) -> ok.
+-spec next_request(binary(), #state{}) -> no_return().
+next_request(Buffer, State=#state{parent=Parent}) ->
+	receive
+		{system, From, Msg} ->
+			cowboy_sys:handle_msg(Msg, From, Parent, ?MODULE, undefined,
+				{State, Buffer});
+		{'EXIT', Parent, Reason} ->
+			terminate(Reason, State)
+	after 0 ->
+		?MODULE:parse_request(Buffer, State, 0)
+	end.
+
+-spec error_terminate(cowboy:http_status(), #state{}) -> no_return().
 error_terminate(Status, State=#state{socket=Socket, transport=Transport,
 		compress=Compress, onresponse=OnResponse}) ->
 	error_terminate(Status, cowboy_req:new(Socket, Transport,
 		undefined, <<"GET">>, <<>>, <<>>, 'HTTP/1.1', [], <<>>,
 		undefined, <<>>, false, Compress, OnResponse), State).
 
--spec error_terminate(cowboy:http_status(), cowboy_req:req(), #state{}) -> ok.
+-spec error_terminate(cowboy:http_status(), cowboy_req:req(), #state{})
+	-> no_return().
 error_terminate(Status, Req, State) ->
 	_ = cowboy_req:reply(Status, Req),
-	terminate(State).
+	terminate(normal, State).
 
--spec terminate(#state{}) -> ok.
-terminate(#state{socket=Socket, transport=Transport}) ->
+-spec terminate(any(), #state{}) -> no_return().
+terminate(Reason, #state{socket=Socket, transport=Transport}) ->
 	Transport:close(Socket),
-	ok.
+	exit(Reason).
```

Comment:
It's been so long I have no context for this but I will happily take the compliment anyway :).

I can see I have an issue with terminate reasons to fix. I also need to rebase.

If you could let me know what your current thinking is on this PR that would be much appreciated. If, or when, you are keen to get things moving here I can put some time to fix it up.